rust/cyclotron-core/src/worker.rs

```diff
@@ -267,22 +309,228 @@ impl Worker {
         // lock after the if check, makes the compiler think the lock is held across
         // the await point.
         {
-            let pending = self.pending.lock().unwrap();
+            let pending = self.running.lock().unwrap();
             if !pending.contains_key(&job_id) {
-                return Err(QueueError::UnknownJobId(job_id));
+                return Err(JobError::UnknownJobId(job_id).into());
             }
         }
 
         dead_letter(&self.pool, job_id, reason).await
     }
 
     /// Passing None here will clear the blob
-    pub fn set_blob(&self, job_id: Uuid, blob: Option<Bytes>) -> Result<(), QueueError> {
-        let mut pending = self.pending.lock().unwrap();
+    pub fn set_blob(&self, job_id: Uuid, blob: Option<Bytes>) -> Result<(), JobError> {
+        let mut pending = self.running.lock().unwrap();
         pending
             .get_mut(&job_id)
-            .ok_or(QueueError::UnknownJobId(job_id))?
+            .ok_or(JobError::UnknownJobId(job_id))?
             .blob = Some(blob);
         Ok(())
     }
 }
+
+// Started by each worker on creation, just loops seeing if the passed batch can be flushed, and
+// if it can, flushing it.
+async fn flush_loop(
+    pool: PgPool,
+    batch: Weak<Mutex<FlushBatch>>,
+    max_buffered: usize,
+    max_bytes: usize,
+    interval: Duration,
+) {
+    loop {
+        let Some(batch) = batch.upgrade() else {
+            // The batch has been dropped, we should exit.
+            break;
+        };
+        // Contemplating sync mutexes on the tree of woe.
+        let mut to_flush = { batch.lock().unwrap().take() };
+        if to_flush.should_flush(max_buffered, max_bytes) {
+            if let Err(e) = to_flush.flush(&pool).await {
+                error!("Error flushing batch: {:?}", e);
+            }
+        }
+        // We can always merge the taken batch back into the pending batch - on successful
+        // flush, the taken batch will be empty, and on failure, we need to re-queue those updates.
```

Comment:
Are there any ordering concerns here?

1. Take batch, start to async flush, it includes an update to Job 1.
2. JS side adds conflicting update to Job 1, expecting last write to win.
3. Async write fails, we re-merge the failed update to Job 1. What happens?

Reply:
In this situation, the second update will be rejected by the worker (it will return a FlushHandle::immediate) because the first release popped the job from the "running" set.

The other way to get multiple updates to a single job into the flush batch is if the worker dequeues a job, stalls, the janitor resets it, the worker releases the job update (no flush yet so the reset isn't observed), calls dequeue and then gets the same job again, and releases a second update before first is flushed. In this case, the lock id changing between dequeues means the first update will be rejected with an invalid lock id, and the second will "win", regardless of what order they're in in the update batch/what order they're literally written in.